### DISCUSSION TITLE
Make it possible to pre-load or autorun py-editor setup

### DISCUSSION THREAD

**Discussion by MarcSkovMadsen** [pyscript/pyscript#2120] [Proposals]
I would like to use py-editor with Panel. The first time you click the run button it takes a long time and the user is not informed why. Only that he/ she has to wait via a loading spinner.

I would like to be able to minimize the time it takes to run by preloading the config and optionally some python code.

You can find a minimal example of what I believe I should be able to do here https://github.com/pyscript/pyscript/issues/2119. The only thing missing is the possibility to add a  `preload` or `autorun` attribute on the `<script type="mpy-editor" config="./config.toml" env="shared" setup>` to trigger a run as soon as possible.

## Alternative

As an alternative is should be described how-to implement this use case if it can already be done via Javascript or Python.

**Comment by WebReflection** [comment]
the editor bootstraps only on run because multiple editors are allowed and the page would be fully busy otherwise but we did, in fact, discussed the ability to autorun.

For the time being, you can add an `id` to your script (not the setup one, any other that is not a setup) and then:

```js
document.getElementById('that-script').target.querySelector('button').click();
```

This would run that specific editor code ASAP so I hope as workaround this can unblock you.

**Reply by MarcSkovMadsen** [reply]
That is almost a working solution for me. Would it be possible (in the future) to trigger the setup only because that is actually What I want to trigger?

**Reply by WebReflection** [reply]
would you like `auto` on `setup` only or you want to `autorun` every single editor? just to be sure I know what to discuss with others, thanks.

**Reply by WebReflection** [reply]
forgot to mention: if there is a setup it takes care of the heavy part eventually shared with all editors later on so I think it's the best thing to do but also, if everyone expects that, I think it might be natural to just bootstrap setup editor while the rest is still shown and editable on the page, yet not executed ... assuming it's clear that the setup node is meant to indeed bootstrap/setup and nothing else. This means less "*yet another attribute to document*" as the amount of attributes and related combinatory logic around should never go "out of hands" for both users and us, maintainers.

**Reply by MarcSkovMadsen** [reply]
Specifically here I was looking for a way to autorun the `setup` in the background before the user hits the `run` button.

But I can easily imagine situations where autorunning other editors would be useful.

**Reply by WebReflection** [reply]
the setup, if present, inevitably runs before any related visual non-seutp editor code, so we can start one step at the time and see if it improves current state already? the rest might be done later on

**Reply by Alex-Monahan** [reply]
Hello! I am also trying to automatically run the editor. I have tried a few ways, but haven't been able to get your snippet to work for me. Where exactly should the document.getElementById snippet you sent be placed? I tried placing it in a script tag of type py, as well as a normal script tag. Maybe you have a full example handy (a complete html snippet would be awesome)?

**Reply by WebReflection** [reply]
`that-script` is the `id` attribute of the editor ... `<script type="mpy-editor" id="that-script">` as example

**Reply by Alex-Monahan** [reply]
Thanks for your reply! I did have the id set and was referencing it. However, I am hitting a concurrency problem I believe - my snippet to click the button is running before the editor loads. How can I be sure to run that script after the editor has loaded?

**Reply by WebReflection** [reply]
you shouldn't need a script to override the Run button, there's a `handleEvent` API to do that you can attach to the script itself, but it's true that due *MutationObserver* asynchronous behavior, it's possible you indeed could have a race there.

```js
(function behavior() {
  const script = document.getElementById('that-script');
  if (!script.target) return setTimeout(behavior);
  // the rest of your code
}());
```

This snippet might help you as it works only once the `target` field has been provided by PyScript.
